spring-beans/src/main/java/org/springframework/beans/factory/groovy/GroovyDynamicElementReader.java

```diff
@@ -0,0 +1,145 @@
+import java.util.Map;
+
+import groovy.lang.Closure;
+import groovy.lang.GroovyObject;
+import groovy.lang.GroovyObjectSupport;
+import groovy.lang.Writable;
+import groovy.xml.StreamingMarkupBuilder;
+import org.w3c.dom.Element;
+
+import org.springframework.beans.factory.config.BeanDefinitionHolder;
+import org.springframework.beans.factory.support.AbstractBeanDefinition;
+import org.springframework.beans.factory.xml.BeanDefinitionParserDelegate;
+
+/**
+ * Used by GroovyBeanDefinitionReader to read a Spring XML namespace expression
+ * in the Groovy DSL.
+ *
+ * @author Jeff Brown
+ * @author Juergen Hoeller
+ * @since 4.0
+ */
+class GroovyDynamicElementReader extends GroovyObjectSupport {
+
+	private final String rootNamespace;
+
+	private final Map<String, String> xmlNamespaces;
+
+	private final BeanDefinitionParserDelegate delegate;
+
+	private final GroovyBeanDefinitionWrapper beanDefinition;
+
+	protected final boolean decorating;
+
+	private boolean callAfterInvocation = true;
+
+
+	public GroovyDynamicElementReader(String namespace, Map<String, String> namespaceMap,
+			BeanDefinitionParserDelegate delegate, GroovyBeanDefinitionWrapper beanDefinition, boolean decorating) {
+		super();
+		this.rootNamespace = namespace;
+		this.xmlNamespaces = namespaceMap;
+		this.delegate = delegate;
+		this.beanDefinition = beanDefinition;
+		this.decorating = decorating;
+	}
+
+
+	@Override
+	public Object invokeMethod(String name, Object obj) {
+		Object[] args = ((Object[])obj);
+		if (name.equals("doCall")) {
+			@SuppressWarnings("unchecked")
+			Closure<Object> callable = (Closure<Object>) args[0];
+			callable.setResolveStrategy(Closure.DELEGATE_FIRST);
+			callable.setDelegate(this);
+			var result = callable.call();
+
+			if (this.callAfterInvocation) {
+				afterInvocation();
+				this.callAfterInvocation = false;
+			}
+			return result;
+		}
+
+		else {
+			StreamingMarkupBuilder builder = new StreamingMarkupBuilder();
+			var myNamespace = this.rootNamespace;
+			var myNamespaces = this.xmlNamespaces;
+
+			Closure<Object> callable = new Closure<>(this) {
+				@Override
+				public Object call(Object... its) {
+					((GroovyObject) getProperty("mkp")).invokeMethod("declareNamespace", new Object[] {myNamespaces});
+					int len = args.length;
+					if (len > 0 && args[len-1] instanceof Closure) {
+						@SuppressWarnings("unchecked")
+						Closure<Object> callable = (Closure<Object>) args[len-1];
+						callable.setResolveStrategy(Closure.DELEGATE_FIRST);
+						callable.setDelegate(builder);
+					}
+					return ((GroovyObject) ((GroovyObject) getDelegate()).getProperty(myNamespace)).invokeMethod(name, args);
+				}
+			};
+
+			callable.setResolveStrategy(Closure.DELEGATE_FIRST);
+			callable.setDelegate(builder);
+			Writable writable = (Writable) builder.bind(callable);
+			var sw = new StringWriter();
+			try {
+				writable.writeTo(sw);
+			}
+			catch (IOException ex) {
+				throw new IllegalStateException(ex);
```

Comment:
Did the Groovy version automatically wrap the `IOException` in an `IllegalStateException`?

Reply:
No, I had to guess. The Groovy version didn't declare throws.